api/envoy/config/listener/v3/listener.proto

```diff
@@ -119,6 +119,40 @@ message Listener {
   // :ref:`FAQ entry <faq_how_to_setup_sni>`.
   repeated FilterChain filter_chains = 3;
 
+  // A list of filter chain matching predicate extension type URLs specifying
+  // the order of selection for the most specific filter chain. At each step,
+  // the predicate is evaluated for each filter chain. If the predicate returns
+  // false, the filter chain is eliminated from subsequent consideration. If
+  // the predicate returns true, but there is a more specific match, it is also
+  // eliminated.
+  //
+  // For example, for SNI ``www.example.com`` the filter chain with the
+  // predicate ```*.example.com``` remains and the filter chain with the
+  // predicate ```*.com``` is eliminated, assuming there is no filter chain
+  // with the exact SNI match. In general, the order of specificity is domain
+  // specific as defined by the predicate extension.
+  //
+  // In case, there is more than one filter chain candidates remaining after
+  // the process completes, the first of the filter chains in the order of
+  // their declaration is selected.
```

Comment:
This is a breaking change. Why do you want to relax the constraint? 

My understanding is that changing match order may impact the last select chain. That's fine.

Reply:
The existing implementation does not guarantee uniqueness, right? E.g. you can have two source port ranges that are overlapping but neither is more specific.

Reply:
It guarantees. An exception will be thrown upon before executing listener update. 

I need to check the source port range, but for other match criteria, e.g. the ip range, we continue to drill down the rest matchers to confirm no 2 chains can be selected

Reply:
Proof
```
  if (!source_ports_map.try_emplace(source_port, filter_chain).second) {
    // If we got here and found already configured branch, then it means that this FilterChainMatch
    // is a duplicate, and that there is some overlap in the repeated fields with already processed
    // FilterChainMatches.
    throw EnvoyException(fmt::format("error adding listener '{}': multiple filter chains with "
                                     "overlapping matching rules are defined",
                                     address_->asString()));
  }
```

Reply:
Mandating that matching predicates form an independent join lattice seems unnecessary. So what if there are two overlapping IP ranges? Maybe it's down-selected further in some other condition.

Reply:
I tend to agree with @kyessenov though I'm wary of changing the implementation in subtle ways for the default. Is it hard to keep the existing constraint for now, at least for predicates that check this type of thing? It seems not too hard?

Reply:
I'll have to look at the implementation in more detail to answer. I am not sure how to detect this statically since you can have a product {sni: ["x", "y"], source_ports: [80, 81]} and {sni: ["x", "z"], source_ports: [79, 80]}. For sni "x" and source port "80" there are two chains but it's not obvious from the definition.

Reply:
This is actually an important usability point. I think the desire to have only one matching chain makes it really hard to use the filter chains. The user expects empty match predicates to match everything, not nothing. So it's up to the user to structure the matches so that only one chain remains in most cases, but we should handle multiple "match-all" chains.